### PR TITLE
dkim: Verify multiple signatures in parallel

### PR DESCRIPTION
**Pull request by foxcpp**
See foxcpp/maddy#7 for related discussion.

### CODE REVIEW

File: dkim/verify.go [hunk 1]
```diff
@@ -111,32 +109,64 @@ func Verify(r io.Reader) ([]*Verification, error) {
...
+		go func() {
+			v, err := verify(h, pr, h[sig.i], sig.v)
+
+			v.Err = err
+			chans[i] <- v
+		}()
+	}
+
+	// This can fail only if CloseWithError is used (we don't do it).
+	io.Copy(io.MultiWriter(writers...), r)
```
Comment (emersion) [inline dkim/verify.go]: `r.Read` can return an error, in which case `io.Copy` will fail.

File: dkim/verify.go [hunk 2]
```diff
@@ -111,32 +109,64 @@ func Verify(r io.Reader) ([]*Verification, error) {
...
-		// Use the bytes.Reader if there is one
-		var r io.Reader = bufr
-		if br != nil {
-			br.Seek(0, io.SeekStart)
-			r = br
-		}
+		// Be careful with loop variables and goroutines.
+		i, sig := i, sig
+
+		chans[i] = make(chan *Verification)
```
Comment (emersion) [inline dkim/verify.go]: We can make the channel non-blocking here, we don't need synchronization.

Comment (foxcpp) [reply dkim/verify.go]: Addressed in e55f212.

File: dkim/verify.go (hunk 1 above)
Comment (foxcpp) [reply dkim/verify.go]: Addressed in e55f212.

**Pull request merged**